File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.py
def solution(phone_book):
    number_map = set()
    phone_book.sort(key = lambda x:len(x))
    for phone_number in phone_book:
        for i in range(1, len(phone_number)):
            sliced_number = phone_number[:i]
            number_map.add(sliced_number)
    # print(number_map)
    for phone_number in phone_book:
        if phone_number in number_map:
            return False
    return True
            

File: 프로그래머스/2/42577. 전화번호 목록/전화번호 목록.java
import java.util.*;

class Solution {
    public boolean solution(String[] phone_book) {
        Arrays.sort(phone_book, Comparator.comparingInt(String::length).reversed()
                                  .thenComparing(Comparator.naturalOrder()));
        
        Set<String> set = new HashSet<>();
        StringBuilder sb = new StringBuilder();
        for(String pNum: phone_book){
            set.add(pNum);
        }
        
        for(String pNum: phone_book){
            for(int i = 0; i < pNum.length() - 1; i++){
                sb.append(pNum.charAt(i));
                if (set.contains(sb.toString())) return false;                
            }
            sb.setLength(0);
        }
        
        return true;
    }
}